File: ClientApp/src/Common/Result.ts
class Error {
    constructor(
        public readonly message: string,
        public readonly key: string | null
    ) { }
}

export class Result {
    constructor(
        public readonly errors: Error[],
    ) { }

    public isSuccess(): boolean {
        if (this.errors.length === 0) return true
        else return false
    }

    public isFail(): boolean {
        if(this.errors.length > 0) return true
        else return false
    }

    public getFirstErrorMessage(): string | null {
        if (!this.errors[0]) return null
        return this.errors[0].message ?? null
    }

    public getFirstErrorKey(): string | null {
        if (!this.errors[0]) return null
        return this.errors[0].key ?? null
    }
}

export function mapToResult(data: any): Result {
    return new Result(data.errors)
}

File: Tasks/Tools/Result.cs
namespace Tools;

public class Error
{
    public String Message { get; }
    public String? Key { get; }

    public Error(String message, String? key = null)
    {
        Message = message; 
        Key = key;
    }
}


public class Result
{
    
    public Error[] Errors { get; }
    public Boolean IsSuccess => Errors.Length == 0;

    public Result()
    {
        Errors = Array.Empty<Error>();
    }

    public Result(String error) 
    {
        Errors = new Error[] { new Error(error) };
    }

    public Result(String key, String error)
    {
        Errors = new Error[] { new Error(error, key) };
    }

    public static Result Success()
    {
        return new Result();
    }

    public static Result Fail(String error)
    {
        return new Result(error);
    }

    public static Result Fail(String key, String error)
    {
        return new Result(key, error);
    }
}
